### ISSUE TITLE
Move to Python 3

### ISSUE THREAD

**Issue by cortesi** [mitmproxy/mitmproxy#326]
Creating this as a catch-all ticket to elicit some discussion of Python3 support, which has become something of an FAQ.

We're going to have to make the jump to Python3 at some point. All our dependencies (last time I checked) were Python 3-ready, so in theory we could make the leap. If we do it, we would do Python 3-only - I don't think there's a benefit to trying to maintain Python 2/3 Frankenstein that gives us the worst of both worlds. 

This leaves us with two issues. First, there's a legacy concern - quite a few people use libmproxy to build things, and I know that I personally would probably have a few weeks of work ahead of me just in converting existing related projects over. 

The second thing has to do with Python 3 and its treatment of Unicode/strings. Mitmproxy is "special", in the sense that it deals with loosely typed, untrusted and possibly intentionally corrupted data flows. It's a boundary layer between a weird world where string encodings can't be assumed, and the higher levels like the UI and so forth, where Python 3 forces us to have unicode. In this sense, it's similar to other programs that deal with on-the-wire data, like web frameworks (though our problem is even worse). Have a look at Armin Ronacher's experience with Python 3 Unicode and Flask/Jinja2/etc.:

http://lucumr.pocoo.org/2014/1/5/unicode-in-2-and-3/

http://lucumr.pocoo.org/2014/5/12/everything-about-unicode/

 I know that this has frustrated some people working on similar programs enough that they've moved away from Python altogether....

Where does this leave us? Well, clearly we clearly can't stay on Python2 in the longer term. The Python3 situation is pretty dire, but has very gradually been getting very slightly better. 3.4 included some improvements, and I think 3.5 will have a few more. The key issues are intrinsic to the language design, though, and will never go away.

**Comment by bradleypeabody** [comment]
fwiw - I'm also a proponent of moving to Python 3 as well and leaving 2 in the dust.  I did a couple of installs a few months ago on CentOS 6.x and had the interesting problem of them only having Python 2.6 available by default. Manually upgrading to 2.7 was a pain and it would have been just as easy to move directly to 3 (probably easier actually).  Legacy support will always be an important issue, but moving forward is just as important (e.g. http://fedoraproject.org/wiki/Changes/Python_3_as_Default https://wiki.ubuntu.com/Python/3 )

P.S. Docker's popularity is also incidentally helping solve this problem - at least for Linux servers.  Supporting a newer Python version for an individual application is a piece of cake using containers/Docker.

**Comment by cortesi** [comment]
A note for readers of this ticket - we're now waiting on PyInstaller to have a mature Python3 branch before we move. It's either that, or begin to roll our own distribution packages for major platforms, which would be a major timesink.

**Comment by bazabaza** [comment]
@cortesi thanks for info. Please keep us up :)

**Comment by elitest** [comment]
I looked into python 2->3 porting over the weekend, and also watched some of the pycon 2014 videos on the topic.

If we were just to move 2.x in the dust, how would it get done?  Code freeze while it gets ported?  Just use python’s 2to3 and fix stuff that breaks?  Or fork and then backport?

I think I’d like to try to make the case for python 2 & 3 compatible code in the same codebase.  The feeling I got was that so much work has gone into python 2.6 and 2.7 around backporting Python 3 stuff that it is often easier to clean up your code than to do a hard break and that python 3 seem to be at a point where there are decent tools that help you write agnostic code.  I had no idea for example that if you do a “python -3 mitmproxy” under recent-ish versions of 2.x, it gives you debug output, about code that it comes across that needs to be addressed for python 3.x.

The other things that seemed to be major roadblocks to porting, were things like CPython extentions, which we don’t do, and not having good testing which we I think we have.  Can Travis help us out in this manner?  I agree that the strings handling vs byte/unicode handling, is probably the biggest problem, but that isn’t going to be solved by dropping 2.x support is it?

references:
http://python3porting.com/
https://www.youtube.com/watch?v=nx0sXSeJbmI
https://www.youtube.com/watch?v=f_6vDi7ywuA

**Comment by cortesi** [comment]
So, there would be a preparatory stage where we get ready for a Python 3 migration by getting rid of Python 2-isms, migrating to backported features, and bring our unit test suite back up to 100%, and so forth. Then there would be a flagfall with a code freeze, and after a few days of effort we'd emerge as a Python 3 only project. I don't think maintaining a 2&3 compatible codebase is worth the effort, unless you're a library has a user constituency in both camps. We're in the relatively luxurious position where we can actually make a complete shift without annoying too many people, and in fact, without most of our users noticing. 

I did a survey recently, and ALL our dependencies seem to be Python 3-ready. The one exception is PyInstaller, a perennial bugbear. Its Python3 branch (https://github.com/pyinstaller/pyinstaller/tree/python3) doesn't seem mature, and is getting less development attention than the main branch. I'm watching to see how that shakes out.

**Comment by mhils** [comment]
I'd propose something along these lines:

1. Make netlib/pathod cross-compatible with Python 2 and 3, mark all workarounds for Python 2. The expected outcome is that the netlib/pathod tests run fine for Python 2 and 3 while mitmproxy is still Python 2 only. This could be done without freezing anything - the required workarounds shouldn't be too severe.
2. Move mitmproxy to Python 3 - we should see whether we can retain Python 2 compatibility for the first adjustments, but we should not be resistant to do a code freeze here.
3. Ditch Python 2 for everything, remove workarounds.

If you want to help, I'd be happy to see some initial work on netlib/pathod that makes it cross-compatible and could be immediately merged into master. My start with the python3 branch was nice to explore the situation, but keeping Pyhton 2 compatibility here for a bit might be useful to make the transition smooth.

**Comment by cortesi** [comment]
That might be worth the effort as an intermediate step. We should definitely wait until we have a plausible path forward with PyInstaller before we move on this - a large fraction of our users rely on this at the moment.

**Comment by JPS46225** [comment]
Hi Max (mhils)

I was just reading your comment about mitmproxy only being supported up to Python 2.7. I was wondering whether this was still the case, or if it had been updated to support the latest Python release?

**Comment by mhils** [comment]
No news yet - I'd be happy to see someone start with netlib. :wink:

**Comment by JPS46225** [comment]
Hi Max

Thanks for replying to my comment.

I've just installed Python 2.7.10, however I'm still having problems installing mitmproxy.

This is the full response from Command Prompt after typing in the command 'pip install mitmproxy'

Collecting mitmproxy
 Using cached mitmproxy-0.12.1.tar.gz
Collecting pyperclip>=1.5.8 (from mitmproxy)
 Using cached pyperclip-1.5.11.zip
Collecting pyasn1>0.1.2 (from mitmproxy)
 Using cached pyasn1-0.1.7.tar.gz
Collecting tornado>=4.0.2 (from mitmproxy)
 Using cached tornado-4.2.tar.gz
Collecting netlib<0.13,>=0.12 (from mitmproxy)
 Using cached netlib-0.12.1.tar.gz
   Complete output from command python setup.py egg_info:
   warning: no files found matching 'OpenSSL\RATIONALE'
   warning: no previously-included files found matching 'leakcheck'
   warning: no previously-included files matching '*.py' found under directory 'leakcheck'
   warning: no previously-included files matching '*.pem' found under directory 'leakcheck'
   warning: no previously-included files matching '*.pyc' found anywhere in distribution
   no previously-included directories found matching 'doc\_build'
   zip_safe flag not set; analyzing archive contents...

   Installed c:\users\admini~1\appdata\local\temp\pip-build-mz8r00\netlib\.eggs\pyopenssl-0.15.1-py2.7.egg
   Searching for cffi
   Reading https://pypi.python.org/simple/cffi/
   Best match: cffi 1.1.0
   Downloading https://pypi.python.org/packages/source/c/cffi/cffi-1.1.0.tar.gz#md5=b58d43a708e757f63a905c6a0d9ecf7a
   Processing cffi-1.1.0.tar.gz
   Writing c:\users\admini~1\appdata\local\temp\easy_install-27xoru\cffi-1.1.0\setup.cfg
   Running cffi-1.1.0\setup.py -q bdist_egg --dist-dir c:\users\admini~1\appdata\local\temp\easy_install-27xoru\cffi-1.1.0\egg-dist-tmp-p10cca
   Traceback (most recent call last):
     File "<string>", line 20, in <module>
     File "c:\users\admini~1\appdata\local\temp\pip-build-mz8r00\netlib\setup.py", line 87, in <module>
       "install": CFFIInstall,
     File "C:\Python27\lib\distutils\core.py", line 111, in setup
       _setup_distribution = dist = klass(attrs)
     File "C:\Python27\lib\site-packages\setuptools\dist.py", line 268, in __init__
       self.fetch_build_eggs(attrs['setup_requires'])
     File "C:\Python27\lib\site-packages\setuptools\dist.py", line 313, in fetch_build_eggs
       replace_conflicting=True,
     File "C:\Python27\lib\site-packages\pkg_resources\__init__.py", line 830, in resolve
       dist = best[req.key] = env.best_match(req, ws, installer)
     File "C:\Python27\lib\site-packages\pkg_resources\__init__.py", line 1075, in best_match
       return self.obtain(req, installer)
     File "C:\Python27\lib\site-packages\pkg_resources\__init__.py", line 1087, in obtain
       return installer(requirement)
     File "C:\Python27\lib\site-packages\setuptools\dist.py", line 380, in fetch_build_egg
       return cmd.easy_install(req)
     File "C:\Python27\lib\site-packages\setuptools\command\easy_install.py", line 629, in easy_install
       return self.install_item(spec, dist.location, tmpdir, deps)
     File "C:\Python27\lib\site-packages\setuptools\command\easy_install.py", line 659, in install_item
       dists = self.install_eggs(spec, download, tmpdir)
     File "C:\Python27\lib\site-packages\setuptools\command\easy_install.py", line 842, in install_eggs
       return self.build_and_install(setup_script, setup_base)
     File "C:\Python27\lib\site-packages\setuptools\command\easy_install.py", line 1070, in build_and_install
       self.run_setup(setup_script, setup_base, args)
     File "C:\Python27\lib\site-packages\setuptools\command\easy_install.py", line 1058, in run_setup
       raise DistutilsError("Setup script exited with %s" % (v.args[0],))
   distutils.errors.DistutilsError: Setup script exited with error: Microsoft Visual C++ 9.0 is required (Unable to find vcvarsall.bat). Get it from http://aka.ms/vcpython27

**Comment by JPS46225** [comment]
Here is the updated command prompt response to 'pip install mitmproxy'

Collecting mitmproxy
  Using cached mitmproxy-0.12.1.tar.gz
Collecting pyperclip>=1.5.8 (from mitmproxy)
  Using cached pyperclip-1.5.11.zip
Collecting pyasn1>0.1.2 (from mitmproxy)
  Using cached pyasn1-0.1.7.tar.gz
Collecting tornado>=4.0.2 (from mitmproxy)
  Using cached tornado-4.2.tar.gz
Collecting netlib<0.13,>=0.12 (from mitmproxy)
  Using cached netlib-0.12.1.tar.gz
    Complete output from command python setup.py egg_info:
    warning: no files found matching 'OpenSSL\RATIONALE'
    warning: no previously-included files found matching 'leakcheck'
    warning: no previously-included files matching '*.py' found under directory 'leakcheck'
    warning: no previously-included files matching '*.pem' found under directory 'leakcheck'
    warning: no previously-included files matching '*.pyc' found anywhere in distribution
    no previously-included directories found matching 'doc\_build'
    zip_safe flag not set; analyzing archive contents...
    
    Installed c:\users\admini~1\appdata\local\temp\pip-build-en7rdx\netlib\.eggs\pyopenssl-0.15.1-py2.7.egg
    Searching for cffi
    Reading https://pypi.python.org/simple/cffi/
    Best match: cffi 1.1.0
    Downloading https://pypi.python.org/packages/source/c/cffi/cffi-1.1.0.tar.gz#md5=b58d43a708e757f63a905c6a0d9ecf7a
    Processing cffi-1.1.0.tar.gz
    Writing c:\users\admini~1\appdata\local\temp\easy_install-2gta8y\cffi-1.1.0\setup.cfg
    Running cffi-1.1.0\setup.py -q bdist_egg --dist-dir c:\users\admini~1\appdata\local\temp\easy_install-2gta8y\cffi-1.1.0\egg-dist-tmp-qh08ku
    _cffi_backend.c
    ffi.c
    prep_cif.c
    types.c
    win32.c
       Creating library build\temp.win32-2.7\Release\c\_cffi_backend.lib and object build\temp.win32-2.7\Release\c\_cffi_backend.exp
    creating c:\users\admini~1\appdata\local\temp\pip-build-en7rdx\netlib\.eggs\cffi-1.1.0-py2.7-win32.egg
    Extracting cffi-1.1.0-py2.7-win32.egg to c:\users\admini~1\appdata\local\temp\pip-build-en7rdx\netlib\.eggs
    
    Installed c:\users\admini~1\appdata\local\temp\pip-build-en7rdx\netlib\.eggs\cffi-1.1.0-py2.7-win32.egg
    Searching for six>=1.5.2
    Reading https://pypi.python.org/simple/six/
    Best match: six 1.9.0
    Downloading https://pypi.python.org/packages/source/s/six/six-1.9.0.tar.gz#md5=476881ef4012262dfc8adc645ee786c4
    Processing six-1.9.0.tar.gz
    Writing c:\users\admini~1\appdata\local\temp\easy_install-plgsfd\six-1.9.0\setup.cfg
    Running six-1.9.0\setup.py -q bdist_egg --dist-dir c:\users\admini~1\appdata\local\temp\easy_install-plgsfd\six-1.9.0\egg-dist-tmp-w5xsci
    no previously-included directories found matching 'documentation\_build'
    zip_safe flag not set; analyzing archive contents...
    six: module references __path__
    creating c:\users\admini~1\appdata\local\temp\pip-build-en7rdx\netlib\.eggs\six-1.9.0-py2.7.egg
    Extracting six-1.9.0-py2.7.egg to c:\users\admini~1\appdata\local\temp\pip-build-en7rdx\netlib\.eggs
    
    Installed c:\users\admini~1\appdata\local\temp\pip-build-en7rdx\netlib\.eggs\six-1.9.0-py2.7.egg
    Searching for cryptography>=0.7
    Reading https://pypi.python.org/simple/cryptography/
    Best match: cryptography 0.9.1
    Downloading https://pypi.python.org/packages/source/c/cryptography/cryptography-0.9.1.tar.gz#md5=6c45d87896f5155f25cdc4d7e0a57526
    Processing cryptography-0.9.1.tar.gz
    Writing c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\setup.cfg
    Running cryptography-0.9.1\setup.py -q bdist_egg --dist-dir c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\egg-dist-tmp-rh7i8c
    _cffi_backend.c
    ffi.c
    prep_cif.c
    types.c
    win32.c
       Creating library build\temp.win32-2.7\Release\c\_cffi_backend.lib and object build\temp.win32-2.7\Release\c\_cffi_backend.exp
    
    Installed c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\.eggs\cffi-1.1.0-py2.7-win32.egg
    Searching for ipaddress
    Reading https://pypi.python.org/simple/ipaddress/
    Best match: ipaddress 1.0.7
    Downloading https://pypi.python.org/packages/source/i/ipaddress/ipaddress-1.0.7.tar.gz#md5=5d9ecf415cced476f7781cf5b9ef70c4
    Processing ipaddress-1.0.7.tar.gz
    Writing c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\temp\easy_install-s1gt3k\ipaddress-1.0.7\setup.cfg
    Running ipaddress-1.0.7\setup.py -q bdist_egg --dist-dir c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\temp\easy_install-s1gt3k\ipaddress-1.0.7\egg-dist-tmp-mxwd0g
    zip_safe flag not set; analyzing archive contents...
    Moving ipaddress-1.0.7-py2.7.egg to c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\.eggs
    
    Installed c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\.eggs\ipaddress-1.0.7-py2.7.egg
    Searching for enum34
    Reading https://pypi.python.org/simple/enum34/
    Best match: enum34 1.0.4
    Downloading https://pypi.python.org/packages/source/e/enum34/enum34-1.0.4.zip#md5=9843e97527f3126c851df7afeb0524b3
    Processing enum34-1.0.4.zip
    Writing c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\temp\easy_install-_3imtl\enum34-1.0.4\setup.cfg
    Running enum34-1.0.4\setup.py -q bdist_egg --dist-dir c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\temp\easy_install-_3imtl\enum34-1.0.4\egg-dist-tmp-ch2bdl
    zip_safe flag not set; analyzing archive contents...
    Moving enum34-1.0.4-py2.7.egg to c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\.eggs
    
    Installed c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\.eggs\enum34-1.0.4-py2.7.egg
    Searching for six>=1.4.1
    Reading https://pypi.python.org/simple/six/
    Best match: six 1.9.0
    Downloading https://pypi.python.org/packages/source/s/six/six-1.9.0.tar.gz#md5=476881ef4012262dfc8adc645ee786c4
    Processing six-1.9.0.tar.gz
    Writing c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\temp\easy_install-sfelkn\six-1.9.0\setup.cfg
    Running six-1.9.0\setup.py -q bdist_egg --dist-dir c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\temp\easy_install-sfelkn\six-1.9.0\egg-dist-tmp-prgjst
    no previously-included directories found matching 'documentation\_build'
    zip_safe flag not set; analyzing archive contents...
    six: module references __path__
    creating c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\.eggs\six-1.9.0-py2.7.egg
    Extracting six-1.9.0-py2.7.egg to c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\.eggs
    
    Installed c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\.eggs\six-1.9.0-py2.7.egg
    Searching for pyasn1
    Reading https://pypi.python.org/simple/pyasn1/
    Best match: pyasn1 0.1.7
    Downloading https://pypi.python.org/packages/2.7/p/pyasn1/pyasn1-0.1.7-py2.7.egg#md5=15f079cabee01402bf86ca8b83356469
    Processing pyasn1-0.1.7-py2.7.egg
    Moving pyasn1-0.1.7-py2.7.egg to c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\.eggs
    
    Installed c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\.eggs\pyasn1-0.1.7-py2.7.egg
    Searching for idna
    Reading https://pypi.python.org/simple/idna/
    Best match: idna 2.0
    Downloading https://pypi.python.org/packages/source/i/idna/idna-2.0.tar.gz#md5=bd17a9d15e755375f48a62c13b25b801
    Processing idna-2.0.tar.gz
    Writing c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\temp\easy_install-hy81xk\idna-2.0\setup.cfg
    Running idna-2.0\setup.py -q bdist_egg --dist-dir c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\temp\easy_install-hy81xk\idna-2.0\egg-dist-tmp-mx45zd
    warning: no previously-included files matching '*.pyc' found under directory 'tools'
    warning: no previously-included files matching '*.pyc' found under directory 'tests'
    zip_safe flag not set; analyzing archive contents...
    Moving idna-2.0-py2.7.egg to c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\.eggs
    
    Installed c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\.eggs\idna-2.0-py2.7.egg
    Searching for pycparser
    Reading https://pypi.python.org/simple/pycparser/
    Best match: pycparser 2.13
    Downloading https://pypi.python.org/packages/source/p/pycparser/pycparser-2.13.tar.gz#md5=e4fe1a2d341b22e25da0d22f034ef32f
    Processing pycparser-2.13.tar.gz
    Writing c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\temp\easy_install-sqxb1e\pycparser-2.13\setup.cfg
    Running pycparser-2.13\setup.py -q bdist_egg --dist-dir c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\temp\easy_install-sqxb1e\pycparser-2.13\egg-dist-tmp-wckizf
    warning: no previously-included files matching 'yacctab.*' found under directory 'tests'
    warning: no previously-included files matching 'lextab.*' found under directory 'tests'
    warning: no previously-included files matching 'yacctab.*' found under directory 'examples'
    warning: no previously-included files matching 'lextab.*' found under directory 'examples'
    zip_safe flag not set; analyzing archive contents...
    Moving pycparser-2.13-py2.7.egg to c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\.eggs
    
    Installed c:\users\admini~1\appdata\local\temp\easy_install-jvjamx\cryptography-0.9.1\.eggs\pycparser-2.13-py2.7.egg
    no previously-included directories found matching 'docs\_build'
    warning: no previously-included files matching '*' found under directory 'vectors'
    _Cryptography_cffi_a269d620xd5c405b7.c
    src\cryptography\hazmat\bindings\__pycache__\_Cryptography_cffi_a269d620xd5c405b7.c(217) : fatal error C1083: Cannot open include file: 'openssl/aes.h': No such file or directory
    Traceback (most recent call last):
      File "<string>", line 20, in <module>
      File "c:\users\admini~1\appdata\local\temp\pip-build-en7rdx\netlib\setup.py", line 87, in <module>
        "install": CFFIInstall,
      File "C:\Python27\lib\distutils\core.py", line 111, in setup
        _setup_distribution = dist = klass(attrs)
      File "C:\Python27\lib\site-packages\setuptools\dist.py", line 268, in __init__
        self.fetch_build_eggs(attrs['setup_requires'])
      File "C:\Python27\lib\site-packages\setuptools\dist.py", line 313, in fetch_build_eggs
        replace_conflicting=True,
      File "C:\Python27\lib\site-packages\pkg_resources\__init__.py", line 830, in resolve
        dist = best[req.key] = env.best_match(req, ws, installer)
      File "C:\Python27\lib\site-packages\pkg_resources\__init__.py", line 1075, in best_match
        return self.obtain(req, installer)
      File "C:\Python27\lib\site-packages\pkg_resources\__init__.py", line 1087, in obtain
        return installer(requirement)
      File "C:\Python27\lib\site-packages\setuptools\dist.py", line 380, in fetch_build_egg
        return cmd.easy_install(req)
      File "C:\Python27\lib\site-packages\setuptools\command\easy_install.py", line 633, in easy_install
        rmtree(tmpdir)
      File "C:\Python27\lib\site-packages\setuptools\command\easy_install.py", line 2208, in rmtree
        rmtree(fullname, ignore_errors, onerror)
      File "C:\Python27\lib\site-packages\setuptools\command\easy_install.py", line 2208, in rmtree
        rmtree(fullname, ignore_errors, onerror)
      File "C:\Python27\lib\site-packages\setuptools\command\easy_install.py", line 2208, in rmtree
        rmtree(fullname, ignore_errors, onerror)
      File "C:\Python27\lib\site-packages\setuptools\command\easy_install.py", line 2213, in rmtree
        onerror(os.remove, fullname, sys.exc_info())
      File "C:\Python27\lib\site-packages\setuptools\command\easy_install.py", line 1609, in auto_chmod
        return func(arg)
    WindowsError: [Error 5] Access is denied: 'c:\\users\\admini~1\\appdata\\local\\temp\\easy_install-jvjamx\\cryptography-0.9.1\\.eggs\\cffi-1.1.0-py2.7-win32.egg\\_cffi_backend.pyd'

**Comment by JPS46225** [comment]
Ok, here is the next updated output of 'pip install mitmproxy' after installing cryptography

Collecting mitmproxy
  Using cached mitmproxy-0.12.1.tar.gz
Collecting pyperclip>=1.5.8 (from mitmproxy)
  Using cached pyperclip-1.5.11.zip
Requirement already satisfied (use --upgrade to upgrade): pyasn1>0.1.2 in c:\python27\lib\site-packages (from mitmproxy)
Collecting tornado>=4.0.2 (from mitmproxy)
  Using cached tornado-4.2.tar.gz
Collecting netlib<0.13,>=0.12 (from mitmproxy)
  Using cached netlib-0.12.1.tar.gz
    Complete output from command python setup.py egg_info:
    warning: no files found matching 'OpenSSL\RATIONALE'
    warning: no previously-included files found matching 'leakcheck'
    warning: no previously-included files matching '*.py' found under directory 'leakcheck'
    warning: no previously-included files matching '*.pem' found under directory 'leakcheck'
    warning: no previously-included files matching '*.pyc' found anywhere in distribution
    no previously-included directories found matching 'doc\_build'
    zip_safe flag not set; analyzing archive contents...
    
    Installed c:\users\admini~1\appdata\local\temp\pip-build-nlejhy\netlib\.eggs\pyopenssl-0.15.1-py2.7.egg
    running egg_info
    creating pip-egg-info\netlib.egg-info
    writing requirements to pip-egg-info\netlib.egg-info\requires.txt
    writing pip-egg-info\netlib.egg-info\PKG-INFO
    writing top-level names to pip-egg-info\netlib.egg-info\top_level.txt
    writing dependency_links to pip-egg-info\netlib.egg-info\dependency_links.txt
    writing manifest file 'pip-egg-info\netlib.egg-info\SOURCES.txt'
    warning: manifest_maker: standard file '-c' not found
    
    running build_ext
    building '_cffi__x88a8fd90x94e38fe2' extension
    creating c:\users\admini~1\appdata\local\temp\pip-build-nlejhy\netlib\netlib\__pycache__\Release
    creating c:\users\admini~1\appdata\local\temp\pip-build-nlejhy\netlib\netlib\__pycache__\Release\netlib
    creating c:\users\admini~1\appdata\local\temp\pip-build-nlejhy\netlib\netlib\__pycache__\Release\netlib\__pycache__
    C:\Users\Administrator\AppData\Local\Programs\Common\Microsoft\Visual C++ for Python\9.0\VC\Bin\cl.exe /c /nologo /Ox /MD /W3 /GS- /DNDEBUG -IC:\Python27\include -IC:\Python27\PC /Tcnetlib\__pycache__\_cffi__x88a8fd90x94e38fe2.c /Foc:\users\admini~1\appdata\local\temp\pip-build-nlejhy\netlib\netlib\__pycache__\Release\netlib\__pycache__\_cffi__x88a8fd90x94e38fe2.obj -w
    cl : Command line warning D9025 : overriding '/W3' with '/w'
    _cffi__x88a8fd90x94e38fe2.c
    netlib\__pycache__\_cffi__x88a8fd90x94e38fe2.c(208) : fatal error C1083: Cannot open include file: 'openssl/rsa.h': No such file or directory
    Traceback (most recent call last):
      File "<string>", line 20, in <module>
      File "c:\users\admini~1\appdata\local\temp\pip-build-nlejhy\netlib\setup.py", line 87, in <module>
        "install": CFFIInstall,
      File "C:\Python27\lib\distutils\core.py", line 151, in setup
        dist.run_commands()
      File "C:\Python27\lib\distutils\dist.py", line 953, in run_commands
        self.run_command(cmd)
      File "C:\Python27\lib\distutils\dist.py", line 972, in run_command
        cmd_obj.run()
      File "<string>", line 15, in replacement_run
      File "C:\Python27\lib\site-packages\setuptools\command\egg_info.py", line 207, in find_sources
        mm.run()
      File "C:\Python27\lib\site-packages\setuptools\command\egg_info.py", line 291, in run
        self.add_defaults()
      File "C:\Python27\lib\site-packages\setuptools\command\egg_info.py", line 320, in add_defaults
        sdist.add_defaults(self)
      File "C:\Python27\lib\site-packages\setuptools\command\sdist.py", line 118, in add_defaults
        build_py = self.get_finalized_command('build_py')
      File "C:\Python27\lib\distutils\cmd.py", line 312, in get_finalized_command
        cmd_obj.ensure_finalized()
      File "C:\Python27\lib\distutils\cmd.py", line 109, in ensure_finalized
        self.finalize_options()
      File "C:\Python27\lib\site-packages\setuptools\command\build_py.py", line 28, in finalize_options
        orig.build_py.finalize_options(self)
      File "C:\Python27\lib\distutils\command\build_py.py", line 46, in finalize_options
        ('force', 'force'))
      File "C:\Python27\lib\distutils\cmd.py", line 298, in set_undefined_options
        src_cmd_obj.ensure_finalized()
      File "C:\Python27\lib\distutils\cmd.py", line 109, in ensure_finalized
        self.finalize_options()
      File "c:\users\admini~1\appdata\local\temp\pip-build-nlejhy\netlib\setup.py", line 27, in finalize_options
        self.distribution.ext_modules = get_ext_modules()
      File "c:\users\admini~1\appdata\local\temp\pip-build-nlejhy\netlib\setup.py", line 20, in get_ext_modules
        from netlib import certffi
      File "netlib\certffi.py", line 20, in <module>
        extra_compile_args=['-w']
      File "C:\Python27\lib\site-packages\cffi\api.py", line 373, in verify
        lib = self.verifier.load_library()
      File "C:\Python27\lib\site-packages\cffi\verifier.py", line 96, in load_library
        self._compile_module()
      File "C:\Python27\lib\site-packages\cffi\verifier.py", line 192, in _compile_module
        outputfilename = ffiplatform.compile(tmpdir, self.get_extension())
      File "C:\Python27\lib\site-packages\cffi\ffiplatform.py", line 38, in compile
        outputfilename = _build(tmpdir, ext)
      File "C:\Python27\lib\site-packages\cffi\ffiplatform.py", line 65, in _build
        raise VerificationError('%s: %s' % (e.__class__.__name__, e))
    cffi.ffiplatform.VerificationError: CompileError: command 'C:\\Users\\Administrator\\AppData\\Local\\Programs\\Common\\Microsoft\\Visual C++ for Python\\9.0\\VC\\Bin\\cl.exe' failed with exit status 2

**Comment by syst3mw0rm** [comment]
@JPS46225 Can you run `python -c "import ssl; print ssl.OPENSSL_VERSION"` and paste the output here?

Also, the issue you are facing is not related to python3 support. I'd suggest to create a separate issue for your problem.

**Comment by JPS46225** [comment]
Hi syst3mw0rm

I will get to pasting output from command that you gave me asap. In regards to creating a new issue, since this isn't related to Python3 support, I actually couldn't find where you can create a new issue. I'll be more than happy to do this once I find out how to do this

**Comment by JPS46225** [comment]
Just found the new issue button, I was looking at this website last time on my iPhone

**Comment by JPS46225** [comment]
Hi syst3mw0rm

As requested, here is the output result of the command that you asked me to enter into Command Prompt:

OpenSSL 1.0.2a 19 Mar 2015

**Comment by JPS46225** [comment]
This is the new issue link for my problem with installing Mitmproxy on Python (Windows)

https://github.com/mitmproxy/mitmproxy/issues/612

**Comment by JPS46225** [comment]
Hi Guys

I was just wondering if there has been any recent progress in regards to the move to Python 3?

**Comment by cattleyavns** [comment]
I think I will try to make mitmproxy work on both Python 2 and 3, well but I need time because mitmproxy is too big, and Twisted have no Python 3 which is the hardest math.

**Comment by JPS46225** [comment]
Hi

Just wondering what you mean by 'twisted with Python native' and also 'less as possible number of third party library', I've only installed what's necessary to get Mitmproxy functioning. Also, is there a way to assign a proxy to a device that doesn't have internal settings to configure proxies, perhaps, through a gateway/router?

**Comment by JPS46225** [comment]
Also, if there is anything else that I can do to improve overall experience, please let me know

**Comment by JPS46225** [comment]
Oh, yes, I see what you mean now by twisted, also I'd love to know how I can learn to do stuff like proxy configuration and other Python modules, if you know of any resources, please share them?

**Comment by cattleyavns** [comment]
Hi @JPS46225, to answer your question:

>> Just wondering what you mean by 'twisted with Python native' and also 'less as possible number of third party library', I've only installed what's necessary to get Mitmproxy functioning. 

Well, we can use socketserver instead Twisted, they are almost the same, but Twisted support something that socketserver don't have, but we can modify socketserver to have those.

>> Also, is there a way to assign a proxy to a device that doesn't have internal settings to configure proxies, perhaps, through a gateway/router?

Like above, socketserver (lower level), HTTPServer or TCPServer (higher).

With socketserver/HTTPServer/TCPServer we can create a proxy server that listen at whatever port we want.

I love socks5 mode for mitmproxy, it can filter/change HTTPS sites, that is a rare feature that almost no socks5 proxy do.

**Comment by JPS46225** [comment]
Hi cattleyavns

You're going to have to assume that I know absolutely nothing about this, so every little detail helps...

I have heard of socket server, but don't get it, would you mind elaborating more on the subject?

**Comment by JPS46225** [comment]
Any tutorials related to the subject, would also help immensely...

**Comment by JPS46225** [comment]
Ok, so, basically what I've known about Mitmproxy is so far, is that you have to tell client device to use the server that's running Mitmproxy, which would be the computer and then tell it the port to use. What I mean by internal settings, is that unlike my iPad, where I can configure a proxy for it to use, the other device which I'm trying this on, doesn't have the option to configure a proxy

**Comment by PidgeyL** [comment]
Is there any estimate (date-wise) of when MitMProxy will be supported in Python 3.x?

**Comment by mhils** [comment]
No - the ROI for porting is still fairly low for us and we currently focus on other features. I'm happy to help anyone who wants to contribute this feature though.

**Comment by riX2000** [comment]
ROI for me was higher than trying to get the python 2 version working. Spent 2 days on that and only like 2 hours porting it to python 3...

**Comment by PidgeyL** [comment]
I also haven't gotten the 2.x version working yet. Do you have the ported version somewhere available?

**Comment by riX2000** [comment]
Sorry, I just edited the installed package, and I'm sure there is several errors and bugs left.
I just ran it with python3 and fixed any errors using 2to3 and some manual fixing until it worked.

The best would be if someone could make a tutorial on how to get it working on a system using python3 as default (Arch in my case). I tried several methods, including AUR, git, pip2 etc.. none worked.

**Comment by PidgeyL** [comment]
So I assume you just downloaded the archive, and imported from its subfolders?

**Comment by mhils** [comment]
Porting mitmproxy to Python 3 unfortunately takes way more time than two
hours. We're really happy for (partial) contributions here, this project is
completely volunteer-driven and this is a good chance to engage yourself.
:-)

**Comment by riX2000** [comment]
@PidgeyL No, I edited the files installed with pip in /usr/lib/site-packages/. Not the recommeneded way of course, just clone the git and go from there, my way was a dirty hack.

@mhils I'm sure, but to just get it working to capture data it took two hours :)

**Comment by PidgeyL** [comment]
@mhils: Would it take long to update just the libmproxy? (which I assume is what @riX2000 did?)

**Comment by riX2000** [comment]
yes, libmproxy (and netlib..)

**Comment by mhils** [comment]
This is still the way to go: https://github.com/mitmproxy/mitmproxy/issues/326#issuecomment-90764563

**Comment by riX2000** [comment]
What about the python3 branch on mitmproxy/netlib? Looks to be about the same fixes I did, is it working?

**Comment by mhils** [comment]
This was a quick attempt to see how far we get with little effort. Directly
porting everything to py3 isn't viable though - py2/3 cross compatibility
is the first goal for netlib.

riX2000 <notifications@github.com> schrieb am Di., 4. Aug. 2015 16:18:

**Comment by riX2000** [comment]
Hmm I was gonig to start porting netlib, but then I noticed one of the reasons the python2 version did not work.. the mitmproxy/dev assumes the python 2 executable is 'python' and equivalently 'pip', which is usually not the case if python 3 is installed. Editing this file to use python2 and pip2 did the trick..
@mhils Maybe add a note about this in the readme or even better a python and pip version check in the dev file.

**Comment by kmike** [comment]
> The only convincing argument for me would be a large external codebase that relies on mitmproxy/libmproxy. 

Scrapy tests use libmproxy; it'd be nice to have Python 2/3 compatible codebase, not just Python 3 rewrite.

> the mitmproxy/dev assumes the python 2 executable is 'python' and equivalently 'pip', which is usually not the case if python 3 is installed.

According to https://www.python.org/dev/peps/pep-0394/ it shouldn't be the case. `python` should be Python 3 only in a Python 3 virtualenv or in distributions which go against this PEP (like Arch Linux).

**Comment by mhils** [comment]
Thanks for the feedback, @kmike. Is it just https://github.com/scrapy/scrapy/blob/master/tests/test_proxy_connect.py or did I miss some parts?

FWIW, I ported netlib a few weeks ago to be cross-comptabile with 2.7 and 3.5. We're not sure what to do with mitmproxy, but we'll take that into consideration.

**Comment by kmike** [comment]
@mhils great to hear netib is ported, thanks! 

> Thanks for the feedback, @kmike. Is it just https://github.com/scrapy/scrapy/blob/master/tests/test_proxy_connect.py or did I miss some parts?

Yes, it is just that. Currently for HTTP proxy tests we're using Twisted.

**Comment by kmike** [comment]
Do you target 3.5 only, not 3.4? Python 3.4 is the default Python 3 in latest Ubuntu LTS, so it'd be nice to support it as well. Are there new features you're using?

**Comment by mhils** [comment]
We're using a few 3.5 features, but nothing that couldn't be fixed easily. I'm personally fine with supporting 3.5 only (which should be the default for Ubuntu 15.10 and then 16.04), but I'd be happy to accept patches that add 3.4 compatibility. :wink:

**Comment by kmike** [comment]
Ok, sounds good :)

**Comment by scone** [comment]
Has anyone begun the porting process? I'm game for working on the port, but don't want to duplicate efforts.

**Comment by mhils** [comment]
I don't think so. If you go for it, please make it  as incremental as you
can (e.g. make the tests for one class or one *.py file pass) and send
partial pull requests as early as possible. :-)

scone <notifications@github.com> schrieb am Mi., 11. Nov. 2015 08:55:

**Comment by ianozsvald** [comment]
+1 for Python 3.5 support, I just tried to install it and was surprised to find `mitmdump` fail with `import thread`. Given we have <4 years to the 2020 sunset date for Python 2.7, I'm surprised to find tools that aren't Python 3.4+ ready (I understand the issues, I talk about this on stage at my monthly PyDataLondon [2,800+ members] meetup to get folk to make the jump). For now I'll downgrade to an old 2.7 install.

**Comment by cortesi** [comment]
@ianozsvald Mitmproxy is a good example of a difficult Python 2-3 transition. If you're interested in the issue - and especially if you feel qualified to speak publicly about it - it would be a good idea to look closely and sympathetically at why we're not there yet. 

- Mitmproxy has lots of dependencies, some of which only recently became Python 3 ready themselves. 
- We're not interested in inelegantly straddling Python 2 and 3 - we'll be moving across to Python3 wholesale. A lot of projects integrate mitmproxy as a library, which means that a move to Python3 will force integrators to either stay with old versions of mitmproxy or to move to Python3 themselves.
- Mitmproxy works at a slightly unusual boundary where Python3's unicode and byte handling becomes a pain point. Armin Rigo has written eloquently about this (though his post is now somewhat outdated): http://lucumr.pocoo.org/2014/5/12/everything-about-unicode/.
- Mitmproxy is distributed as a compiled binary for all major platforms. We rely on pyinstaller for this, but pyinstaller has only supported Python3 in a stable release since last October. Pyinstaller has been a huge pain-point for us, and I'm waiting for the Python3 support to mature a bit before the transition.

I hope you will be less surprised next time you find a large, complicated tool still stuck on 2.7. The bug tracker is not the place to discuss this (feel free to drop me an email), but I've come to consider the way the Python2-3 transition was handled to be one of the greatest technical blunders of the last decade. Long after mitmproxy itself has shifted to Python3, the vast bulk of production Python code will still be on 2.7. We'd better make peace with that. 

If you'd like to help things along for mitmproxy itself, please join us on the Slack channel to discuss, and then to roll up your sleeves and pitch in. ;)

**Comment by ianozsvald** [comment]
Hi Aldo. Sorry, I wasn't trying to be unsympathetic. I had seen the discussion around a wholesale shift from Py2 to Py3 (this seems like a sensible approach if feasible), I understand the legacy difficulties with longer running users. I've followed and have been involved in the discussions around the move of projects like numpy, matplotlib and django (years back), scrapy (just now feasible via Twisted's updates), PySpark and lots of smaller projects. I've ported a few very small projects (several hours each to successful PR, a few involving web scraping/byte issues), nothing on the scale of mitmproxy.

I'm very aware of the str/unicode/byte pain point and I've followed Armin's posts over the years. I spend almost all of my time in the data science world where almost all the major ports were completed years ago (with dual version support). A part of my outspoken-ness at public meetups is to try to get the slower moving users to upgrade (which is totally feasible in the data science world now) such that the developers have an easier time supporting just the one Python 3 platform.

Re. your request that I pitch in - you're totally right that I should if I wanted the port to move faster, however I won't. mitmproxy was a test for a side project and the Py2.7 version did what I needed. Elsewhere I do put plenty of time into the Py community (right now I'm co-chair for our 3rd annual PyDataLondon O/S conference, I'm also co-chair for our monthly 200+ user meetup), so I hope you'll understand. Sorry to cause noise.

**Comment by Kriechi** [comment]
Just a quick update:

Thanks to the great work of @dufferzafar as part of this years GSoC, we are getting close to running mitmproxy in Python 3.5. As of a few days ago, pathod is now fully compatible with Python 2.7 and 3.5. netlib was already py3-compatible, thanks to the great efforts by @mhils a couple of months ago.
This means all dependencies for mitmproxy are able to run in Python 2 and 3.

We expect to migrate the remaining mitmproxy code to Python 3 within the next weeks - so stay tuned!

**Comment by ianozsvald** [comment]
Hello again, it is great to hear about the Python 3 porting progress. A while back I wondered if I could use my PyDataLondon audience to gather some useful data for a couple of projects (notably yours, h5py, dask and a couple of others). Last week I queried my London audience, 13% of my 3,400+ members responded to 4 emailed survey prompts. Here's a summary of the results:
http://ianozsvald.com/2016/06/20/results-for-which-version-of-python-2vs3-do-london-data-scientists-use/

In short - my London Data Science audience predominantly use Python 2.7 at work, Python 3.4+ is dominant outside of work and I hypothesise that Python 3.x to be dominant roughly this time next year. Python <=2.6 and <=3.3 are hardly used by any of this audience. Possibly these figures are useful in future discussions you might have about which Python versions need support into the future. Cheers, Ian.

**Comment by mhils** [comment]
Quick update: We managed to build Python 3.5 binaries for pathod and we only have a couple of modules left for mitmproxy:

- [ ] test/mitmproxy/test_dump.py (#1299)
- [ ] test/mitmproxy/test_examples.py
- [ ] test/mitmproxy/test_flow.py
- [ ] test/mitmproxy/test_flow_format_compat.py (#1318)
- [ ] test/mitmproxy/test_protocol_http2.py (#1307)
- [ ] test/mitmproxy/test_web_master.py

**Comment by mhils** [comment]
While there is still a bit to do, thanks to @dufferzafar's great GSoC work all tests pass now on Python 3! :tada: :snake:

**Comment by cortesi** [comment]
The Python2/3 straddle is now complete. Whatever remains should be split up, and attached to the milestone for 1.0, which will be our first Python3-only release.
